File: Mobile APP-flutter-dart/lib/services/level_service.dart
import 'dart:convert';
import 'package:flutter/services.dart' show rootBundle;
import '../models/level_model.dart';

class LevelService {
  static const String _levelsJsonPath = 'assets/data/levels.json';
  
  // Singleton pattern
  static final LevelService _instance = LevelService._internal();
  factory LevelService() => _instance;
  LevelService._internal();
  
  List<Level> _cachedLevels = [];
  bool _isInitialized = false;
  
  // Initialiser le service
  Future<void> initialize() async {
    if (_isInitialized) return;
    
    try {
      _cachedLevels = await loadLevelsFromJson();
      _isInitialized = true;
    } catch (e) {
      print('Erreur lors de l\'initialisation du LevelService: $e');
      _cachedLevels = Level.getSampleLevels();
      _isInitialized = true;
    }
  }
  
  // Charger les niveaux depuis le fichier JSON
  Future<List<Level>> loadLevelsFromJson() async {
    try {
      final String jsonString = await rootBundle.loadString(_levelsJsonPath);
      final List<dynamic> jsonData = json.decode(jsonString);
      return jsonData.map((levelJson) => Level.fromJson(levelJson)).toList();
    } catch (e) {
      print('Erreur lors du chargement des niveaux depuis JSON: $e');
      // En cas d'erreur, utiliser les niveaux par défaut
      return Level.getSampleLevels();
    }
  }
  
  // Obtenir tous les niveaux
  Future<List<Level>> getAllLevels() async {
    await initialize();
    return _cachedLevels;
  }
  
  // Obtenir un niveau par son ID
  Future<Level?> getLevelById(int id) async {
    await initialize();
    try {
      return _cachedLevels.firstWhere((level) => level.id == id);
    } catch (e) {
      print('Niveau $id non trouvé: $e');
      return null;
    }
  }
  
  // Ajouter un nouveau niveau
  Future<void> addLevel(Level newLevel) async {
    await initialize();
    
    try {
      // Charger les niveaux existants
      final levels = await loadLevelsFromJson();
      
      // Vérifier si l'ID existe déjà
      if (levels.any((level) => level.id == newLevel.id)) {
        throw Exception('Un niveau avec cet ID existe déjà');
      }
      
      // Ajouter le nouveau niveau
      levels.add(newLevel);
      
      // Convertir en JSON
      final jsonData = levels.map((level) => level.toJson()).toList();
      final jsonString = json.encode(jsonData);
      
      // Écrire dans le fichier
      // Note: Cette opération nécessite des permissions d'écriture
      // qui ne sont pas disponibles dans une application Flutter standard
      // Cette méthode serait à implémenter avec un backend ou une solution de stockage
      print('Nouveau niveau ajouté (simulation): ${newLevel.name}');
      
      // Mettre à jour le cache
      _cachedLevels = levels;
    } catch (e) {
      print('Erreur lors de l\'ajout d\'un niveau: $e');
    }
  }
  
  // Mettre à jour un niveau existant
  Future<void> updateLevel(Level updatedLevel) async {
    await initialize();
    
    try {
      // Charger les niveaux existants
      final levels = await loadLevelsFromJson();
      
      // Trouver l'index du niveau à mettre à jour
      final index = levels.indexWhere((level) => level.id == updatedLevel.id);
      if (index == -1) {
        throw Exception('Niveau non trouvé');
      }
      
      // Mettre à jour le niveau
      levels[index] = updatedLevel;
      
      // Convertir en JSON
      final jsonData = levels.map((level) => level.toJson()).toList();
      final jsonString = json.encode(jsonData);
      
      // Écrire dans le fichier (simulation)
      print('Niveau mis à jour (simulation): ${updatedLevel.name}');
      
      // Mettre à jour le cache
      _cachedLevels = levels;
    } catch (e) {
      print('Erreur lors de la mise à jour d\'un niveau: $e');
    }
  }
  
  // Supprimer un niveau
  Future<void> deleteLevel(int levelId) async {
    await initialize();
    
    try {
      // Charger les niveaux existants
      final levels = await loadLevelsFromJson();
      
      // Filtrer le niveau à supprimer
      final filteredLevels = levels.where((level) => level.id != levelId).toList();
      
      if (levels.length == filteredLevels.length) {
        throw Exception('Niveau non trouvé');
      }
      
      // Convertir en JSON
      final jsonData = filteredLevels.map((level) => level.toJson()).toList();
      final jsonString = json.encode(jsonData);
      
      // Écrire dans le fichier (simulation)
      print('Niveau supprimé (simulation): ID $levelId');
      
      // Mettre à jour le cache
      _cachedLevels = filteredLevels;
    } catch (e) {
      print('Erreur lors de la suppression d\'un niveau: $e');
    }
  }
} 

File: Backend-nodejs/services/levelService.js
const { Level } = require('../models');
const fs = require('fs');
const path = require('path');

class LevelService {
  constructor() {
    this.levelsJsonPath = path.join(__dirname, '../data/levels.json');
  }

  // Charger les niveaux depuis le fichier JSON
  loadLevelsFromJson() {
    try {
      if (!fs.existsSync(this.levelsJsonPath)) {
        console.error('Fichier levels.json non trouvé:', this.levelsJsonPath);
        return [];
      }
      const jsonData = fs.readFileSync(this.levelsJsonPath, 'utf8');
      return JSON.parse(jsonData);
    } catch (error) {
      console.error('Erreur lors du chargement des niveaux depuis JSON:', error);
      return [];
    }
  }

  // Sauvegarder les niveaux dans le fichier JSON
  saveLevelsToJson(levels) {
    try {
      // Créer le dossier data s'il n'existe pas
      const dataDir = path.dirname(this.levelsJsonPath);
      if (!fs.existsSync(dataDir)) {
        fs.mkdirSync(dataDir, { recursive: true });
      }
      fs.writeFileSync(this.levelsJsonPath, JSON.stringify(levels, null, 2), 'utf8');
      return true;
    } catch (error) {
      console.error('Erreur lors de la sauvegarde des niveaux dans JSON:', error);
      return false;
    }
  }

  // Synchroniser les niveaux depuis le fichier JSON vers la base de données
  async syncLevelsFromJson() {
    try {
      const levelsData = this.loadLevelsFromJson();
      if (levelsData.length === 0) {
        throw new Error('Aucun niveau trouvé dans le fichier JSON');
      }

      // Vérifier si les niveaux existent déjà
      const existingLevels = await Level.count();
      
      if (existingLevels === 0) {
        // Créer tous les niveaux
        await Level.bulkCreate(levelsData);
        console.log(`${levelsData.length} niveaux créés depuis le fichier JSON`);
      } else {
        // Mettre à jour les niveaux existants et créer les nouveaux
        for (const levelData of levelsData) {
          const [level, created] = await Level.upsert(levelData, {
            returning: true,
          });
          if (created) {
            console.log(`Niveau ${levelData.id} créé`);
          } else {
            console.log(`Niveau ${levelData.id} mis à jour`);
          }
        }
      }

      return levelsData.length;
    } catch (error) {
      console.error('Erreur lors de la synchronisation des niveaux:', error);
      throw error;
    }
  }

  // Synchroniser les niveaux depuis la base de données vers le fichier JSON
  async syncLevelsToJson() {
    try {
      const levels = await Level.findAll({
        order: [['id', 'ASC']],
      });

      const levelsData = levels.map(level => ({
        id: level.id,
        name: level.name,
        instruction: level.instruction,
        code: level.code,
        codeLength: level.codeLength,
        pointsReward: level.pointsReward,
        isLocked: level.isLocked,
        timeLimit: level.timeLimit,
        additionalHints: level.additionalHints,
        hintCost: level.hintCost,
      }));

      this.saveLevelsToJson(levelsData);
      console.log(`${levelsData.length} niveaux sauvegardés dans le fichier JSON`);
      return levelsData.length;
    } catch (error) {
      console.error('Erreur lors de la synchronisation des niveaux vers JSON:', error);
      throw error;
    }
  }

  // Obtenir tous les niveaux depuis la base de données
  async getAllLevels() {
    try {
      const levels = await Level.findAll({
        order: [['id', 'ASC']],
      });
      return levels;
    } catch (error) {
      console.error('Erreur lors de la récupération des niveaux:', error);
      throw error;
    }
  }

  // Obtenir un niveau par son ID
  async getLevelById(id) {
    try {
      const level = await Level.findByPk(id);
      return level;
    } catch (error) {
      console.error('Erreur lors de la récupération du niveau:', error);
      throw error;
    }
  }

  // Créer ou mettre à jour un niveau
  async upsertLevel(levelData) {
    try {
      const [level, created] = await Level.upsert(levelData, {
        returning: true,
      });
      return { level, created };
    } catch (error) {
      console.error('Erreur lors de la création/mise à jour du niveau:', error);
      throw error;
    }
  }

  // Supprimer un niveau
  async deleteLevel(id) {
    try {
      const deleted = await Level.destroy({
        where: { id },
      });
      return deleted > 0;
    } catch (error) {
      console.error('Erreur lors de la suppression du niveau:', error);
      throw error;
    }
  }

  // Réinitialiser les niveaux (supprimer tous et recréer depuis JSON)
  async resetLevels() {
    try {
      // Supprimer tous les niveaux
      await Level.destroy({ where: {} });
      
      // Recréer depuis le fichier JSON
      return await this.syncLevelsFromJson();
    } catch (error) {
      console.error('Erreur lors de la réinitialisation des niveaux:', error);
      throw error;
    }
  }
}

module.exports = new LevelService();

